### PR TITLE
Fix fuzzy spatial weights

### PR DESCRIPTION
**Pull request by cgsandford**
This PR re-writes the `SpatiallyVaryingWeightsFromMask` plugin to work with normalised weights.  This allows models to be given higher weightings in the area near the edges of the nowcast mask, where fuzzy weights have been applied, which avoids a hard boundary in the blended output.

Updates have been made to unit test output values and I can explain every one on request.  New acceptance test data have been generated.  Met Office reviewers are directed to a notebook with plots demonstrating that this works correctly and fixes the problems seen in the alpha suite output.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/blending/spatial_weights.py [hunk 1]
```diff
@@ -307,21 +303,28 @@ def process(self, cube_to_collapse, one_dimensional_weights_cube, blend_coord):
...
+            # Set masked weights to zero
+            weights.data = np.where(template_cube.data.mask, 0, weights.data)
+        else:
+            message = "Expected masked input to SpatiallyVaryingWeightsFromMask"
+            warnings.warn(message)
+            return weights
+
+        weights = self._normalise_initial_weights(weights)
+        weights, rescaled = self._rescale_masked_weights(weights)
+        self._rescale_unmasked_weights(weights, rescaled)
```
Comment (BelligerG) [inline improver/blending/spatial_weights.py:328]: When we use masked arrays in a normal (i.e. non-fuzzy-weight) environment, we can sometimes get away with not normalising here, as the weights get normalised within the np.ma operations (e.g. mean). I think, however, because of the operations here, we do need to normalise initially, as you have done.

I wondered if you should put a comment in just saying why the normalisation has to happen here rather than relying on np.ma operations.

File: improver/blending/spatial_weights.py [hunk 2]
```diff
@@ -289,15 +150,150 @@ def create_template_slice(cube_to_collapse, blend_coord):
...
+        is_rescaled = iris.cube.CubeList()
+        rescaled_weights = iris.cube.CubeList()
+        for weights_slice in weights.slices_over(self.blend_coord):
+            weights_nonzero = np.where(weights_slice.data > 0, 1, 0)
+            if np.all(weights_nonzero == 1):
+                # if there are no masked points in this slice, keep current weights
+                # and mark as unchanged (not rescaled)
+                rescaled_weights.append(weights_slice)
+                is_rescaled.append(
+                    weights_slice.copy(data=np.zeros(weights_slice.shape, dtype=int))
```
Comment (BelligerG) [inline improver/blending/spatial_weights.py:229]: This could be a `bool` dtype I think, as the only allowed values are 0 or 1.

File: improver/blending/spatial_weights.py [hunk 3]
```diff
@@ -289,15 +150,150 @@ def create_template_slice(cube_to_collapse, blend_coord):
...
+                # multiply existing weights by fuzzy scaling factor
+                rescaled_weights_data = np.multiply(
+                    weights_slice.data, fuzzy_factor
+                ).astype(FLOAT_DTYPE)
+                rescaled_weights.append(weights_slice.copy(data=rescaled_weights_data))
+
+                # identify spatial points where weights have been rescaled
+                is_rescaled_data = np.where(
+                    rescaled_weights_data != weights_orig, 1, 0
+                ).astype(int)
```
Comment (BelligerG) [inline improver/blending/spatial_weights.py:254]: Same as above, I think this could be `bool`

File: improver/blending/spatial_weights.py [hunk 4]
```diff
@@ -289,15 +150,150 @@ def create_template_slice(cube_to_collapse, blend_coord):
...
+        Args:
+            weights (iris.cube.Cube):
+                Cube of weights to which fuzzy smoothing has been applied to any
+                masked slices
+            is_rescaled (iris.cube.Cube):
+                Cube matching weights.shape, with value of 1 where masked weights
+                have been rescaled, and 0 where they are unchanged.
+        """
+        rescaled_data = np.multiply(weights.data, is_rescaled.data)
+        unscaled_data = np.multiply(weights.data, (1 - is_rescaled.data))
```
Comment (BelligerG) [inline improver/blending/spatial_weights.py:277]: If `is_rescaled` is a `bool` array, you can use the bitwise operator here `~` instead of `1-`

**Review by BelligerG — CHANGES_REQUESTED** [review]
Looks good, a few comments from me though, the largest change would come from re-using already existing broadcasting code.

File: improver/blending/spatial_weights.py [hunk 5]
```diff
@@ -289,15 +150,150 @@ def create_template_slice(cube_to_collapse, blend_coord):
...
+        unscaled_sum = np.sum(unscaled_data, axis=self.blend_axis)
+        required_sum = 1.0 - np.sum(rescaled_data, axis=self.blend_axis)
+        normalisation_factor = np.where(
+            unscaled_sum > 0, np.divide(required_sum, unscaled_sum), 0
+        )
+        normalised_weights = (
+            np.multiply(unscaled_data, normalisation_factor) + rescaled_data
+        )
+        weights.data = normalised_weights.astype(FLOAT_DTYPE)
```
Comment (BelligerG) [inline improver/blending/spatial_weights.py:286]: Is the casting required? `weights.data` should always be in the correct dtype as this is a private function

**Comment by TomekTrzeciak** [comment]
we should look at memory measurements of these changes to gauge their impact

File: improver/blending/spatial_weights.py (hunk 5 above)
Comment (cgsandford) [reply improver/blending/spatial_weights.py:286]: `np.multiply` escalates the datatype, for reasons that I don't understand (since all the inputs should have a datatype of 32-bit precision).

File: improver/blending/spatial_weights.py (hunk 4 above)
Comment (cgsandford) [reply improver/blending/spatial_weights.py:277]: Will that multiply correctly?  The point is I want zeros vs original values.  I suppose I could use a `np.where` with a boolean array.

Comment (BelligerG) [reply improver/blending/spatial_weights.py:277]: It should do :-)

File: improver/blending/spatial_weights.py (hunk 5 above)
Comment (BelligerG) [reply improver/blending/spatial_weights.py:286]: That's very strange, it should only scale to the highest dtype, but I have to be honest, I've never checked it with np.multiply

**Comment by BelligerG** [comment]
> we should look at memory measurements of these changes to gauge their impact

We should, having a quick scan it doesn't look like it'll have changed much from the previous implementation, but having some numbers to prove/disprove would be useful

File: improver/blending/spatial_weights.py [hunk 6]
```diff
@@ -307,21 +303,28 @@ def process(self, cube_to_collapse, one_dimensional_weights_cube, blend_coord):
...
-        weights_from_mask = self.create_initial_weights_from_mask(template_cube)
-        weights_from_mask = self.smooth_initial_weights(weights_from_mask)
-        final_weights = self.multiply_weights(
-            weights_from_mask, one_dimensional_weights_cube, blend_coord
-        )
-        return final_weights
+        template_cube = self._create_template_slice(cube_to_collapse)
+        (self.blend_axis,) = template_cube.coord_dims(self.blend_coord)
+
+        weights = self._broadcast_weights(template_cube, one_dimensional_weights_cube)
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/blending/spatial_weights.py:316]: I prefer my implementation, if I'm honest.  It's cleaner, only does what it needs to, and minimises interdependencies between plugins.

File: improver/blending/spatial_weights.py (hunk 1 above)
Comment (cgsandford) [reply improver/blending/spatial_weights.py:328]: I could elaborate in the "process" docstring - that's probably the best place to put this.

File: improver/blending/spatial_weights.py (hunk 4 above)
Comment (cgsandford) [reply improver/blending/spatial_weights.py:277]: OK, will give this a go.

File: improver/blending/spatial_weights.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/blending/spatial_weights.py:316]: I mean, I could make it a utility and import it in both places if you like?

File: improver/blending/spatial_weights.py (hunk 2 above)
Comment (cgsandford) [reply improver/blending/spatial_weights.py]: Done/

File: improver/blending/spatial_weights.py (hunk 3 above)
Comment (cgsandford) [reply improver/blending/spatial_weights.py]: Done.

File: improver/blending/spatial_weights.py (hunk 5 above)
Comment (cgsandford) [reply improver/blending/spatial_weights.py:284]: Turns out there's actually an argument to `np.multiply` to enforce a result dtype.  However given what's happening in this function, it's actually tidier just to cast the result at the end.

File: improver/blending/spatial_weights.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (TomekTrzeciak) [reply improver/blending/spatial_weights.py:322]: > I prefer my implementation, if I'm honest. It's cleaner, only does what it needs to, and minimises interdependencies between plugins.

Except it destroys the array view on 1D data (memory increase) and goes via intermediate cubes + merge (more memory waste). Feel free to factor out `WeightedBlendAcrossWholeDimension.shape_weights` into a separate utility if it helps, but keep the implementation.

[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/blending/spatial_weights.py:322]: OK, will factor out.

I assume the memory cleverness is down to the `iris.util.broadcast_to` call?  This I could easily put in the `_broadcast_weights` method.  But the `shape_weights` method should be at least two different methods at first glance - one to sort out the dimensions overall, the other to do the actual broadcasting - and I really don't want to be calling a ton of stuff I don't need (this causes problems down the line).

[the remark this answers is not in the document]
Comment (BelligerG) [reply improver/blending/spatial_weights.py:322]: Definitely one to factor out :-) then any optimisations applied to one will automatically be applied to the other.

With regard to the memory increase, anytime you alter a broadcasted array you save a new array in memory, instead of a (basically free) view of the array, so changing the dtype, adding a mask etc... Because of the way the weights need to be adjusted and masked at different x,y positions for each time used, I couldn't easily see a way to do this without creating an intermediate cube.

File: improver_tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py [hunk 7]
```diff
@@ -749,41 +320,35 @@ def test_no_fuzziness_with_one_dimensional_weights(self):
...
-            self.one_dimensional_weights_cube,
-            "forecast_reference_time",
+        result = self.plugin.process(
+            self.cube_to_collapse, self.one_dimensional_weights_cube,
         )
         self.assertArrayAlmostEqual(result.data, expected_result)
         self.assertEqual(result.metadata, self.cube_to_collapse.metadata)
 
     @ManageWarnings(ignored_messages=["Collapsing a non-contiguous coordinate."])
     def test_fuzziness_with_one_dimensional_weights(self):
```
Comment (fionaRust) [inline improver_tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py:339]: This test made me think a bit because of the different but overlapping masks in different slices of the array. Do you think we need a test to show what happens if one slice has masked data in rather than two? The second renormalisation should redistribute weight to both of the other slices evenly for the points that have been scaled with the fuzzy weights.

Comment (fionaRust) [reply improver_tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py:339]: Another edge case to consider. What happens if you only had two slices, and both had missing data in gridboxes either side of a point that was valid in both slices. Then this middle point would be reduced in both slices and the renormalisation wouldn't change it (I think), so you'd end up with weights that didn't add up to 1. I think this would be ok because the weighted blending code where the weights are used would renormalise and reweight them again?
e.g. 
[[--, 0.5, 0.5], [0.5, 0.5, --]] might result in weights [[0, 0.25, 1][0, 0.25, 1]]

**Review by fionaRust — COMMENTED** [review]
A minor comment and a few questions about possible extra unit tests

File: improver_tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py (hunk 7 above)
Comment (cgsandford) [reply improver_tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py:339]: The second edge case, as discussed, I think is dealt with fine due to the initial normalisation.  For the case of 2 unmasked and 1 masked slice, I've now added a unit test to cover this.

File: improver/blending/spatial_weights.py [hunk 8]
```diff
@@ -289,39 +150,187 @@ def create_template_slice(cube_to_collapse, blend_coord):
...
-                vary along the blend_coord, and not along any other dimensions
-                on the cube.
+                vary along the blend and spatial coordinates, and not along
+                any other dimensions on the cube.
             one_dimensional_weights_cube (iris.cube.Cube):
                 A cube containing a single dimension coordinate with the same
                 name given blend_coord. This cube contains 1D weights
                 that will be applied along the blend_coord but need
                 adjusting spatially based on missing data.
-            blend_coord (str):
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply improver/blending/spatial_weights.py:310]: Done

**Review by fionaRust — DISMISSED** [review]
I'm happy with this PR now. Thanks Caroline

**Comment by fionaRust** [comment]
@BelligerG I've now approved this PR. Are you happy with it now and are the green team happy for it to be merged?

File: improver/blending/spatial_weights.py [hunk 9]
```diff
@@ -289,39 +152,156 @@ def create_template_slice(cube_to_collapse, blend_coord):
...
+        happens after the application of fuzzy smoothing near mask boundaries.
+        Modifies weights cube in place.
+
+        Args:
+            weights (iris.cube.Cube):
+                3D weights containing zeros for masked points, but before
+                fuzzy smoothing
+        """
+        weights_sum = np.sum(weights.data, axis=self.blend_axis)
+        weights_sum = np.broadcast_to(weights_sum, weights.shape)
```
Comment (BelligerG) [inline improver/blending/spatial_weights.py:167]: I don't think this line is required, it essentially uses vectorisation to do the divide (which can be done without the memory view).

**Review by BelligerG — CHANGES_REQUESTED** [review]
One very minor change for me, then I'm happy to approve.

**Review by BelligerG — APPROVED** [review]
Looks good to me!

File: improver/blending/spatial_weights.py [hunk 10]
```diff
@@ -244,33 +108,33 @@ def create_template_slice(cube_to_collapse, blend_coord):
...
-        original_dim_coords = [coord.name() for coord in cube_to_collapse.dim_coords]
-        # Slice over relevant coords.
+        original_dim_coords = get_dim_coord_names(cube_to_collapse)
+        # Slice over required coords
         x_coord = cube_to_collapse.coord(axis="x").name()
         y_coord = cube_to_collapse.coord(axis="y").name()
-        coords_to_slice_over = [blend_coord, y_coord, x_coord]
+        coords_to_slice_over = [self.blend_coord, y_coord, x_coord]
+        # If the dimensions are already as required, return cube unchanged
+        if original_dim_coords == coords_to_slice_over:
```
Comment (fionaRust) [inline improver/blending/spatial_weights.py:133]: There's a codecov flag saying this line if statement isn't tested. 
See unit test improver_tests/blending/spatial_weights/test_SpatiallyVaryingWeightsFromMask.py::Test__create_template_slice::test_basic_no_change

After printing some stuff I think it's cos your get_dim coords function classes air_temperature as a dim coord. This means this air temp coord gets removed from teh result cube, even though the test doesn't fail.

Comment (cgsandford) [reply improver/blending/spatial_weights.py:133]: Yes, that's because air temperature IS a dim coord... I'd forgotten that our cubes would have that.
I'll set up a test with just the three coordinates on the input cube.

**Review by fionaRust — APPROVED** [review]
